### ISSUE TITLE
Scanner issue: mis-matched "_id" and "file" records in DB lead to endless healthcheck loops

### ISSUE THREAD

**Issue by jackmaninov** [HaveAGitGat/Tdarr#821]
**Describe the bug**
I've been cleaning up my video library, and after upgrading my docker images to the post 2.00.22 versions (currently on 2.00.24) and clearing my libraries to do a fresh run, I seem to have stumbled on an issue with the file scanner. I am running on a Synology NAS (DSM 7.2-64570 Update 1) under docker, pulling from haveagitgat/tdarr:latest. The bug seems to be associating different files with each other in Tdarr/DB2/FileJSONDB, giving json files with "_id" and "file" fields pointing to different video files. 

I wrote a quick regex to identify these "bad" db entries (checks if _id and file match), then if I run a `watch` in the FileJSONDB dir I can see how many "bad" db entries there are, e.g.:

```
export REGEX='"_id":"(.*?)","file":"\1"'
export FINDER="/usr/bin/grep -v -nE '$REGEX' *.json |wc -l"
watch "$FINDER && echo -n "/" && ls | wc -l"
```
which outputs something like:
```
12
/128
```

For my small movies folder, I end up after a scan with something around 12/184 bad entries, however I have to re-"Scan (Find New)" a few times in order for all the files to be picked up. Every time the bad entries point to different files, so it is not related to the files themselves This is reproducible for me when restarting tdarr_server, clearing all files in my library, killing other services on my NAS to reduce load, etc etc. If I scan all my libraries the percentage of bad entries isn't so high, maybe 260/1800, so I suspect the bug gets triggered when the scanner moves from one directory to the next (my other library contains TV episodes, so there are more files per directory).

With verbose logging off, I see Tdarr_Server throwing several identical TypeErrors: 

```
[2023-08-06T11:05:42.751] [ERROR] Tdarr_Server - TypeError: Cannot read properties of undefined (reading '_id')
    at /app/Tdarr_Server/srcug/fileScanners.js:1:21508
    at _0x375211 (/app/Tdarr_Server/srcug/fileScanners.js:1:4424)
    at Object.next (/app/Tdarr_Server/srcug/fileScanners.js:1:2629)
    at _0x93a3ae (/app/Tdarr_Server/srcug/fileScanners.js:1:1615)
```

however there are fewer TypeErrors thrown than "bad" db entries created (e.g. 6 are thrown on a scan giving 12/184 bad db entries), and they are thrown by the final queue builder that runs at the end of the scan (earlier ones do not throw errors). With verbose logging on I don't see anything interesting, just the normal healthcheck tagging completing successfully.

I first noticed these bad db entries because when the file in the "_id" field gets brought up for a health check, Tdarr will just loop endlessly health checking it (successfully) over and over. This results in high load on my NAS and Tdarr being unable to complete processing of my libraries. Notably the transcode queue does not get stuck on the bad entries, but all files in this library have already been transcoded to hevc, matching both my and the default plugin stacks.

I can eventually work through a library through a combination of re-scanning and manually inspecting the "bad" files in the Search section of the web frontend, but this only works well on small subsets of my library.

**To Reproduce**
1. Start with a fresh tdarr_server (I delete everything in /docker/tdarr/server with my config)
2. Add my movies library with all default settings (source: /volume2/Media/Movies, transcode cache: /volume2/Media/downloads/transcode)
3. Start my watch process described above in directory /docker/tdarr/server/Tdarr/DB2/FileJSONDB
4. Select Scan (Fresh)
5. watch "bad" db entries tick up with files scanned
6. Enable some CPU health check workers
7. Eventually the health check queue gets stuck, looping over and over on files with "bad" db entries

**Expected behavior**
Files are correctly imported into the DB in one pass, health checks do not loop 

**Screenshots**
A bad file with looping health checks (they run every few seconds)
![image](https://github.com/HaveAGitGat/Tdarr/assets/1815854/a5b4647b-3183-4cc0-9775-1aa599dc27fb)

the corresponding "bad" db entry (movie title scrubbed):
`
2f2f9de3fa2de6239cb376d63ebd0842ce82a0ce.json:1:{"_id":"/volume2/Media/Movies/SOME MOVIE (2001)/Extras/SOME MOVIE Extras - BLA BLA BLA.mkv","file":"/volume2/Media/Movies/SOME OTHER MOVIE (2012)/SOME OTHER MOVIE (2012) {imdb-BLAH} [Remux-2160p][HDR10][AC3 5.1][EN][10bit][HEVC][EN+FR+ES]-LEGi0N.mkv","DB":"PYITjU9EY","footprintId":"-O0YOxnKnDz","hasClosedCaptions":false,"container":"mkv","scannerReads":{"ffProbeRead":"success"},"ffProbeData":{"format":{},"streams":[{"codec_name":"hevc","codec_type":"video","tags":{"language":"eng","title":"English"},"width":3840,"height":2160},{"codec_name":"ac3","codec_type":"audio","bit_rate":"448000","channels":6,"tags":{"language":"eng","title":"English / DTS-HD Master Audio / 7.1 / 48 kHz / 3 263 kbps / 24-bit (DTS Core: 5.1 / 48 kHz / 1509 kbps / 24-bit)"}},{"codec_name":"ac3","codec_type":"audio","bit_rate":"448000","channels":6,"tags":{"language":"eng","title":"English"}},{"codec_name":"ac3","codec_type":"audio","bit_rate":"640000","channels":6,"tags":{"language":"eng","title":"English"}},{"codec_name":"hdmv_pgs_subtitle","codec_type":"subtitle","tags":{"language":"eng","title":"English"},"duration":"6075.141000"},{"codec_name":"hdmv_pgs_subtitle","codec_type":"subtitle","tags":{"language":"eng","title":"English"},"duration":"6075.141000"},{"codec_name":"hdmv_pgs_subtitle","codec_type":"subtitle","tags":{"language":"fre","title":"French"},"duration":"6075.141000"},{"codec_name":"hdmv_pgs_subtitle","codec_type":"subtitle","tags":{"language":"spa","title":"Spanish"},"duration":"6075.141000"}]},"file_size":36881.41325759888,"video_resolution":"4KUHD","fileMedium":"video","video_codec_name":"hevc","audio_codec_name":"","lastPluginDetails":"none","createdAt":1691313238970,"bit_rate":50926179,"duration":6075,"statSync":{"mtimeMs":1351728000000,"ctimeMs":1689937804320.6682},"HealthCheck":"Queued","TranscodeDecisionMaker":"Queued","lastHealthCheckDate":0,"holdUntil":0,"lastTranscodeDate":0,"bumped":false,"history":"","oldSize":0,"newSize":0,"videoStreamIndex":0}
`

**Please provide the following information:**
Tdarr_Server_Config.json:
```
{
  "serverPort": "8266",
  "webUIPort": "8265",
  "serverIP": "0.0.0.0",
  "handbrakePath": "",
  "ffmpegPath": "",
  "logLevel": "INFO",
  "mkvpropeditPath": "",
  "ccextractorPath": "",
  "openBrowser": true,
  "cronPluginUpdate": ""
}
```

My docker config json (all media mounted on /volume2/Media to match other apps in my stack)
```
{
   "CapAdd" : [],
   "CapDrop" : [],
   "cmd" : "",
   "cpu_priority" : 50,
   "enable_publish_all_ports" : false,
   "enable_restart_policy" : true,
   "enable_service_portal" : null,
   "enabled" : false,
   "env_variables" : [
      {
         "key" : "PATH",
         "value" : "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
      },
      {
         "key" : "HOME",
         "value" : "/home/Tdarr"
      },
      {
         "key" : "LANGUAGE",
         "value" : "en_US.UTF-8"
      },
      {
         "key" : "LANG",
         "value" : "en_US.UTF-8"
      },
      {
         "key" : "TERM",
         "value" : "xterm"
      },
      {
         "key" : "S6_CMD_WAIT_FOR_SERVICES_MAXTIME",
         "value" : "0"
      },
      {
         "key" : "S6_VERBOSITY",
         "value" : "1"
      },
      {
         "key" : "S6_STAGE2_HOOK",
         "value" : "/docker-mods"
      },
      {
         "key" : "LIBVA_DRIVERS_PATH",
         "value" : "/usr/lib/x86_64-linux-gnu/dri"
      },
      {
         "key" : "LD_LIBRARY_PATH",
         "value" : "/usr/lib/x86_64-linux-gnu"
      },
      {
         "key" : "NVIDIA_DRIVER_CAPABILITIES",
         "value" : "all"
      },
      {
         "key" : "NVIDIA_VISIBLE_DEVICES",
         "value" : "all"
      },
      {
         "key" : "HANDBRAKE",
         "value" : "1.6.1"
      },
      {
         "key" : "WEB_UI_PORT",
         "value" : "8265"
      },
      {
         "key" : "SERVER_PORT",
         "value" : "8266"
      },
      {
         "key" : "NODE_PORT",
         "value" : "8267"
      },
      {
         "key" : "PUID",
         "value" : "1026"
      },
      {
         "key" : "PGID",
         "value" : "100"
      },
      {
         "key" : "UMASK",
         "value" : "002"
      },
      {
         "key" : "TZ",
         "value" : "Europe/SomePlace"
      },
      {
         "key" : "MODULE_LINK",
         "value" : ""
      },
      {
         "key" : "NODE_LINK",
         "value" : ""
      },
      {
         "key" : "BASE_LINK",
         "value" : "https://tdarrs.s3.us-west-000.backblazeb2.com/dev/versions/2.00.24"
      },
      {
         "key" : "inContainer",
         "value" : "true"
      },
      {
         "key" : "serverIP",
         "value" : "0.0.0.0"
      }
   ],
   "exporting" : false,
   "id" : "642efc9bd003ea9809ad33d052bff31fd8c8e9a93b449b60d5977714e13494e5",
   "image" : "haveagitgat/tdarr:latest",
   "is_ddsm" : false,
   "is_package" : false,
   "labels" : {
      "build_version" : "Linuxserver.io version:- 127ce7ef-ls120 Build-date:- 2023-06-06T23:27:21+00:00",
      "maintainer" : "TheLamer",
      "org.opencontainers.image.authors" : "linuxserver.io",
      "org.opencontainers.image.created" : "2023-06-06T23:27:21+00:00",
      "org.opencontainers.image.description" : "baseimage-ubuntu image by linuxserver.io",
      "org.opencontainers.image.documentation" : "https://docs.linuxserver.io/images/docker-baseimage-ubuntu",
      "org.opencontainers.image.licenses" : "GPL-3.0-only",
      "org.opencontainers.image.ref.name" : "1f7f55be0549e05f0d0690b5088dcebcd97bc7e4",
      "org.opencontainers.image.revision" : "1f7f55be0549e05f0d0690b5088dcebcd97bc7e4",
      "org.opencontainers.image.source" : "https://github.com/linuxserver/docker-baseimage-ubuntu",
      "org.opencontainers.image.title" : "Baseimage-ubuntu",
      "org.opencontainers.image.url" : "https://github.com/linuxserver/docker-baseimage-ubuntu/packages",
      "org.opencontainers.image.vendor" : "linuxserver.io",
      "org.opencontainers.image.version" : "127ce7ef-ls120"
   },
   "links" : [],
   "memory_limit" : 0,
   "name" : "tdarr-server",
   "network" : [
      {
         "driver" : "bridge",
         "name" : "bridge"
      }
   ],
   "network_mode" : "bridge",
   "port_bindings" : [
      {
         "container_port" : 8265,
         "host_port" : 8265,
         "type" : "tcp"
      },
      {
         "container_port" : 8266,
         "host_port" : 8266,
         "type" : "tcp"
      }
   ],
   "privileged" : false,
   "services" : null,
   "shortcut" : {
      "enable_shortcut" : false,
      "enable_status_page" : false,
      "enable_web_page" : false,
      "web_page_url" : ""
   },
   "use_host_network" : false,
   "version" : 2,
   "volume_bindings" : [
      {
         "host_volume_file" : "/Media",
         "is_directory" : true,
         "mount_point" : "/volume2/Media/",
         "type" : "rw"
      },
      {
         "host_volume_file" : "/docker/tdarr/configs",
         "is_directory" : true,
         "mount_point" : "/app/configs",
         "type" : "rw"
      },
      {
         "host_volume_file" : "/docker/tdarr/logs",
         "is_directory" : true,
         "mount_point" : "/app/logs",
         "type" : "rw"
      },
      {
         "host_volume_file" : "/docker/tdarr/server",
         "is_directory" : true,
         "mount_point" : "/app/server",
         "type" : "rw"
      }
   ]
}
```

- Log files  [can be found in /app/logs/ when using Docker or in the /logs folder next to Tdarr_Updater if not using Docker]
Relevant fragment given above 

- Worker error: no errors show up on the worker, except connection errors if the server load goes too high

 - OS: Linux (Synology DSM 7.2)
 - Browser: Brave latest
 - Version: 2.00.24 pulled from docker image haveagitgat/tdarr:latest

**Additional context**
This setup has been working in the past with older images, and I have already confirmed all permissions are good.

**Comment by HaveAGitGat** [comment]
Hi thanks for the details. I just tried to replicate this using a few hundred files and didn't encounter this issue:
![image](https://github.com/HaveAGitGat/Tdarr/assets/43864057/02c147a1-55ce-4ea0-a605-a66aa492e10d)


Used this js script to try and find problematic files:

```
const url = 'http://0.0.0.0:8266/api/v2/cruddb';

const data = {
  data: {
    collection: 'FileJSONDB',
    mode: 'getAll',
  },
};

const run = async () => {
  const headers = {
    'Content-Type': 'application/json',
  };

  const errFiles = [];

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(data),
    });

    const files = await response.json();

    console.log(files.length);

    for (let i = 0; i < files.length; i += 1) {
      const file = files[i];
      if (file._id !== file.file) {
        errFiles.push(file);
      }
    }

    console.log(errFiles);
  } catch (error) {
    console.error(error);
  }
};

run();

```

Are you able to run that? You can press F12 and paste it into a Chrome snippet to run it in browser:

![image](https://github.com/HaveAGitGat/Tdarr/assets/43864057/d8774f38-9b88-4bce-b2ab-e686ccd62bc2)

**Comment by HaveAGitGat** [comment]
Also try running this js script on your files and see if you get the same result:

```
const fs = require('fs');

const run = async () => {
  const errFiles = [];

  try {
    const folder = '/docker/tdarr/server/Tdarr/DB2/FileJSONDB';

    const files = fs.readdirSync(folder);

    console.log(`files.length: ${files.length}`);

    for (let i = 0; i < files.length; i += 1) {
      const file = files[i];

      const json = fs.readFileSync(`${folder}/${file}`, 'utf8');

      if (json._id !== json.file) {
        errFiles.push(json);
      }
    }

    console.log({
      errFiles,
      totalFiles: files.length,
      errFilesLength: errFiles.length,
      correctFiles: files.length - errFiles.length
    });
  } catch (err) {
    console.error(err);
  }
};

run();
```

Save it locally as test.js and then run `node test.js`. Make sure to have node.js installed.

**Comment by jackmaninov** [comment]
Ok, so I've started off with a fresh server config, I add my movies folder and hit Scan (Fresh). I end up with 118 files found (of the I think 185 there are if I hit rescan a bunch).

The `node test.js` script run at the console agrees with my regex showing 16/118 bad files:

```
...
  totalFiles: 118,
  errFilesLength: 16,
  correctFiles: 102
}
```

while the in-browser script shows more:

```
...
    "totalFiles": 118,
    "errFilesLength": 33,
    "correctFiles": 85
}
```

In addition, both methods produce errFiles that do not show up in the other. I don't want to post the results publicly so I'll jump on the discord.

**Comment by jackmaninov** [comment]
Tdarr_Server_Log.txt for that run. Note that 184 files were sent for extraction.

```
[2023-08-06T23:04:22.139] [INFO] Tdarr_Server - Reading DB:>>>>>>>>>>(100%):0:SettingsGlobalJSONDB::Reading done
[2023-08-06T23:04:22.141] [INFO] Tdarr_Server - Reading DB:>>>>>>>>>>(100%):0:LibrarySettingsJSONDB::Reading done
[2023-08-06T23:04:22.141] [INFO] Tdarr_Server - Reading DB:>>>>>>>>>>(100%):0:NodeJSONDB::Reading done
[2023-08-06T23:04:22.142] [INFO] Tdarr_Server - Reading DB:>>>>>>>>>>(100%):0:StatisticsJSONDB::Reading done
[2023-08-06T23:04:22.311] [INFO] Tdarr_Server - /app/configs/Tdarr_Server_Config.json
[2023-08-06T23:04:22.419] [INFO] Tdarr_Server - {
  serverPort: '8266',
  webUIPort: '8265',
  serverIP: '0.0.0.0',
  handbrakePath: '',
  ffmpegPath: '',
  logLevel: 'INFO',
  mkvpropeditPath: '',
  ccextractorPath: '',
  openBrowser: true,
  cronPluginUpdate: ''
}
[2023-08-06T23:04:23.801] [INFO] Tdarr_Server - version: 2.00.24
[2023-08-06T23:04:23.801] [INFO] Tdarr_Server - platform_arch_isdocker: linux_x64_docker_true
[2023-08-06T23:04:23.802] [INFO] Tdarr_Server - Preparing environment
[2023-08-06T23:04:23.802] [INFO] Tdarr_Server - Tdarr started.
[2023-08-06T23:04:23.802] [INFO] Tdarr_Server - Tdarr documents folder:/app/server
[2023-08-06T23:04:23.803] [INFO] Tdarr_Server - Checking directories
[2023-08-06T23:04:23.812] [INFO] Tdarr_Server - Tdarr_Server listening at http://0.0.0.0:8266
[2023-08-06T23:04:23.813] [TRACE] Tdarr_Server - 
    --------------------------------------------------------------------------
    --------------------------------------------------------------------------
    
    Tdarr_WebUI running at http://localhost:8265/#/
    
    --------------------------------------------------------------------------
    --------------------------------------------------------------------------
      
[2023-08-06T23:04:24.298] [INFO] Tdarr_Server - Node few-fish [ID=57rnfyYEd] [Address=::ffff:192.168.50.177] [PID=10028] registered
[2023-08-06T23:04:24.812] [INFO] Tdarr_Server - Checking binary paths
[2023-08-06T23:04:24.840] [INFO] Tdarr_Server - Running binary tests
[2023-08-06T23:04:25.989] [INFO] Tdarr_Server - ---------------Binary tests start----------------
[2023-08-06T23:04:25.990] [INFO] Tdarr_Server - handbrakePath:HandBrakeCLI
[2023-08-06T23:04:25.990] [INFO] Tdarr_Server - ffmpegPath:tdarr-ffmpeg
[2023-08-06T23:04:25.990] [INFO] Tdarr_Server - mkvpropedit:mkvpropedit
[2023-08-06T23:04:25.991] [INFO] Tdarr_Server - Binary test 1: handbrakePath working
[2023-08-06T23:04:25.991] [INFO] Tdarr_Server - Binary test 2: ffmpegPath working
[2023-08-06T23:04:25.991] [INFO] Tdarr_Server - Binary test 3: mkvpropeditPath working
[2023-08-06T23:04:25.991] [INFO] Tdarr_Server - ---------------Binary tests end-------------------
[2023-08-06T23:04:25.991] [INFO] Tdarr_Server - Running scanner tests
[2023-08-06T23:04:25.992] [INFO] Tdarr_Server - ---------------Scanner tests start----------------
[2023-08-06T23:04:25.992] [INFO] Tdarr_Server - ffprobePath:/app/Tdarr_Server/node_modules/@ffprobe-installer/linux-x64/ffprobe
[2023-08-06T23:04:25.992] [INFO] Tdarr_Server - ccextractorPath:ccextractor
[2023-08-06T23:04:25.993] [INFO] Tdarr_Server - Scanner test: Scanning file start
[2023-08-06T23:04:28.347] [INFO] Tdarr_Server - Scanner test: Scanning file end
[2023-08-06T23:04:28.348] [INFO] Tdarr_Server - Scanner test 1: FFprobe working
[2023-08-06T23:04:28.348] [INFO] Tdarr_Server - Scanner test 2: Exiftool working
[2023-08-06T23:04:28.348] [INFO] Tdarr_Server - Scanner test 3: Mediainfo working
[2023-08-06T23:04:28.348] [INFO] Tdarr_Server - Scanner test 4: CCExtractor working
[2023-08-06T23:04:28.348] [INFO] Tdarr_Server - ---------------Scanner tests end----------------
[2023-08-06T23:04:28.350] [INFO] Tdarr_Server - Database has finished initialising
[2023-08-06T23:04:28.571] [INFO] Tdarr_Server - Updating plugins
[2023-08-06T23:04:28.651] [INFO] Tdarr_Server - Job report history size is within limit. Limit:10240 MiB, Size:0 MiB
[2023-08-06T23:04:28.652] [INFO] Tdarr_Server - Cleaning empty job report folders
[2023-08-06T23:04:29.124] [INFO] Tdarr_Server - Cloning plugins
[2023-08-06T23:04:29.646] [INFO] Tdarr_Server - Cache cleaner start
[2023-08-06T23:04:29.649] [INFO] Tdarr_Server - Cache cleaner end
[2023-08-06T23:04:32.009] [INFO] Tdarr_Server - Finished downloading plugins!
[2023-08-06T23:04:40.236] [INFO] Tdarr_Server - [11116ms]Plugin update finished
[2023-08-06T23:05:32.427] [INFO] Tdarr_Server - 6EDWv9TDY Prep started
[2023-08-06T23:05:32.428] [INFO] Tdarr_Server - Commencing fresh file scan.
[2023-08-06T23:05:32.642] [INFO] Tdarr_Server - [215ms] 6EDWv9TDY Prep finished
[2023-08-06T23:05:32.643] [INFO] Tdarr_Server - Scanner 6EDWv9TDY launched
[2023-08-06T23:05:32.875] [INFO] Tdarr_Server - Online
[2023-08-06T23:05:33.937] [INFO] Tdarr_Server - Sending 184 files for extraction
[2023-08-06T23:05:38.585] [INFO] Tdarr_Server - Finished in 5710ms
[2023-08-06T23:05:38.586] [INFO] Tdarr_Server - Scanner 6EDWv9TDY:Finished
[2023-08-06T23:05:38.587] [WARN] Tdarr_Server - Exit approved. Closing process.
[2023-08-06T23:05:39.597] [INFO] Tdarr_Server - 6EDWv9TDY File scanner exited
```

**Comment by jackmaninov** [comment]
After running "Scan (Find New)" a total of four times, the in-browser result shows 184 correct files no errFiles, while on the disk the console result is 184 total files, 15 errFiles.

**Comment by jackmaninov** [comment]
Went back to some older versions. and confirmed that the breaking change happened between 2.00.21 and 2.00.22. If I run the same test on 2.00.21, I get no errFiles, and the scanner finds all the files after one "fresh" scan and one "find new" scan. 2.00.22 shows the bug same as 2.00.24.

**Comment by jackmaninov** [comment]
However, re-running a fresh scan on my entire library using 2.00.21 gives no errFiles, but still gives the TypeError noted in my initial report, as well as a SyntaxError that I've seen once or twice before but cannot reproduce. This occurs at the end of I suppose "extraction" phase that occurs after the initial file scanner run (after "sending X files for extraction")

```
[2023-08-07T10:29:22.230] [INFO] Tdarr_Server - verbose:This exact file does not exist in DB. Adding
[2023-08-07T10:29:22.536] [INFO] Tdarr_Server - verbose:Adding file to DB: /volume2/Media/TV/SOME SHOW (1986)/Season 02/SOME SHOW (1986) - S02E07 - SOME SHOW TITLE [SDTV][AC3 2.0][EN][8bit][h265][EN].mkv
[2023-08-07T10:29:22.539] [ERROR] Tdarr_Server - SyntaxError: Unexpected token u in JSON at position 0
    at JSON.parse (<anonymous>)
    at /snapshot/be/srcug/fileScanners.js:1:20465
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:24.395] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:24.396] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:24.874] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:25.170] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:25.990] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:27.230] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:28.309] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:29.457] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:29.464] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:29.746] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:32.137] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:32.175] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)
[2023-08-07T10:29:32.181] [ERROR] Tdarr_Server - TypeError: Cannot read property '_id' of undefined
    at /snapshot/be/srcug/fileScanners.js:1:21656
    at _0x49f1a0 (/snapshot/be/srcug/fileScanners.js:1:4364)
    at Object.next (/snapshot/be/srcug/fileScanners.js:1:2625)
    at _0x49ab42 (/snapshot/be/srcug/fileScanners.js:1:1645)

**Comment by HaveAGitGat** [comment]
Are you able to try this: `tdarr_acc:dev_2.00.25_2023_08_07T13_36_58z`, made some small changes which might fix it. 

Also I don't think your watch command is quite right:
```
export REGEX='"_id":"(.*?)","file":"\1"'
export FINDER="/usr/bin/grep -v -nE '$REGEX' *.json |wc -l"
watch "$FINDER && echo -n "/" && ls | wc -l"
```

If I run my script on the files then it shows 0 issues but if I run your watch command it shows 3 issues so it seems it's potentially picking up some other `file` or `_id` strings

**Comment by jackmaninov** [comment]
Ok, with that dev version I get no more errFiles, and I get 150/183 of my files imported on the first fresh scan. I noticed that there was an extra temporary json.2971815680 file that hangs around for a bit and was giving me a total of 184 before. Certainly useable now, as the first folder refresh picks up the remainder of the files.

I still see the SyntaxError and multiple TypeError's during the extraction phase that I was seeing on 2.00.21, but as I say this appears to be a different issue:

```
[2023-08-07T21:17:18.838] [INFO] Tdarr_Server - verbose:Adding file to DB: /volume2/Media/Movies/SOME MOVIE (2003)/SOME MOVIE (2003) {imdb-BLAH} [DVD][MP3 2.0][10bit][h265]-ArtSubs.mkv
[2023-08-07T21:17:18.839] [ERROR] Tdarr_Server - SyntaxError: Unexpected token u in JSON at position 0
    at JSON.parse (<anonymous>)
    at /app/Tdarr_Server/srcug/fileScanners.js:1:20406
    at _0x26a288 (/app/Tdarr_Server/srcug/fileScanners.js:1:6704)
    at Object.next (/app/Tdarr_Server/srcug/fileScanners.js:1:4950)
    at _0x284f93 (/app/Tdarr_Server/srcug/fileScanners.js:1:3931)
[2023-08-07T21:17:18.845] [ERROR] Tdarr_Server - TypeError: Cannot read properties of undefined (reading '_id')
    at /app/Tdarr_Server/srcug/fileScanners.js:1:21590
    at _0x26a288 (/app/Tdarr_Server/srcug/fileScanners.js:1:6704)
    at Object.next (/app/Tdarr_Server/srcug/fileScanners.js:1:4950)
    at _0x284f93 (/app/Tdarr_Server/srcug/fileScanners.js:1:3931)
[2023-08-07T21:17:19.768] [ERROR] Tdarr_Server - TypeError: Cannot read properties of undefined (reading '_id')
    at /app/Tdarr_Server/srcug/fileScanners.js:1:21590
    at _0x26a288 (/app/Tdarr_Server/srcug/fileScanners.js:1:6704)
    at Object.next (/app/Tdarr_Server/srcug/fileScanners.js:1:4950)
    at _0x284f93 (/app/Tdarr_Server/srcug/fileScanners.js:1:3931)
[2023-08-07T21:17:20.861] [INFO] Tdarr_Server - verbose:[6761ms] Queue builder finished. 132 files processed
[2023-08-07T21:17:21.861] [INFO] Tdarr_Server - Queue builder launched
[2023-08-07T21:17:22.489] [ERROR] Tdarr_Server - TypeError: Cannot read properties of undefined (reading '_id')
    at /app/Tdarr_Server/srcug/fileScanners.js:1:21590
    at _0x26a288 (/app/Tdarr_Server/srcug/fileScanners.js:1:6704)
    at Object.next (/app/Tdarr_Server/srcug/fileScanners.js:1:4950)
    at _0x284f93 (/app/Tdarr_Server/srcug/fileScanners.js:1:3931)
[2023-08-07T21:17:24.990] [INFO] Tdarr_Server - verbose:[3128ms] Queue builder finished. 150 files processed
[2023-08-07T21:17:25.991] [INFO] Tdarr_Server - Queue builder launched
[2023-08-07T21:17:27.736] [INFO] Tdarr_Server - verbose:[1744ms] Queue builder finished. 150 files processed
```

I didn't test that regex extensively, it was just to try to quickly debug the problem.

**Comment by HaveAGitGat** [comment]
Ok that’s good that it’s working. It’s strange because all I did going from 2.00.21 to 2.00.22 regarding this was make a small function async which seems to have caused some race condition. Is your Tdarr app data on an HDD? Guess it might be to do with that.

And yeah the other issue is unrelated, seems  maybe ffprobe can’t scan the file properly. Are you able to try scanning it manually with ffprobe?

**Comment by jackmaninov** [comment]
Yeah, I'm running on an HDD, although there's enough RAM in this NAS that everything should be getting cached as I run the tests over and over. Maybe the difference in speed when it has to go to read from the disk causes the race condition.

Anyways, I checked all my files with ffprobe and there were no problems. But after moving a few files around (dupes, subtitles, etc) I can no longer repeatably reproduce the SyntaxError/TypeErrors, but my fresh scans still always stop out at around 145/180 files, so I still suspect there's still a race condition going on. 

I'm working on a shell wrapper for ffprobe to see if I can capture exactly where the problem is, but no luck so far.

**Comment by HaveAGitGat** [comment]
Have you checked that the other files don't have containers that are excluded in the library settings? Else you could try enabling verbose logs on the logs tab and run the scan, it gives much more detail.

I was able to reproduce your original issue by putting my tdarr data on an old HDD I have and making sure there was lots of disk activity before running a scan, so investigating what causes that.

In the meantime I undid the change that caused that issue and released 2.00.25 so that side of things should be working fine now. Feel free to open a new issue if you discover anything regarding those files.
